### ISSUE TITLE
Stacked barplot for UpSetR ?

### ISSUE THREAD

**Issue by Grokitach** [hms-dbmi/UpSetR#59]
Hello,

Is it possible to make stacked barplots based on different labels using UpSetR classes intersections ? I'm using a data frame where each item (row) is defined by:
i) an intersection among the different sets (coded by 0 and 1) used by UpSetR
ii) a text label that could be used by the stacked barplot

Using the example dataset "movies", the goal of this plot would be to analyse both the intersection counts of differents genres, and the number of movies released at each date for each intersection.

I'm working with genomic data comprising regions of interest (one by row) belonging to various intersections among 7 sets (different conditions, correspondig to genres in the movies dataset) and each region is also defined by its location on the genome (genic, intergenic, promoter, corresponding to ReleaseDate in the movies dataset). I would thus be interested by combining UpSetR classes intersection representation with a stacked barplot that would take into account the localization of each item in the various genomic features (genic, intergenic, promoters).

Is it possible to do it ? Using the examples in the help of UpSetR, I did not see how to do it.

Any help would be greatly appreciated.

**Comment by jananiravi** [comment]
Yes, stacked barplots with attributes from other columns would be most helpful. There is a way to achieve 'pseudo-stacking' when active=T, but since the counts start from 0, interpretation becomes difficult. Also, forcing multiple queries to mimic stacking doesn't seem to work with the T or active=F option (with the triangles). An additional toggle parameter to make the counts cumulative could suffice for the time-being serving as a makeshift option.

Is there any other way to get upset plots to talk to ggplot or specify attribute plots of our choice? If so, stacked plots could be added as attribute plots, if not substitute the main one. Any suggestions on work-arounds would help too. Thanks!

![image](https://user-images.githubusercontent.com/8397074/29126998-cb0befd6-7ced-11e7-91ce-b3e79edb249a.png)

**Comment by psudmant** [comment]
I love this idea - what would be very simple ( I THINK???) and useful would be to either make dodged bar plots, OR, to be able to overlay a point or a label on each bar. In my case I'm trying to show the "expected" overlap vs the "observed" overlap if that makes sense.

**Comment by aphalo** [comment]
Yes, indeed, this would be very useful.

**Comment by 1cuigx** [comment]
I have been trying to make a stacked barplot for a while, then realized that there is no "easy" way to create it. Since the issue has been raised for some time, was there any updates on this? It would be really nice if this could be added in the next version.

**Comment by acpguedes** [comment]
@jananiravi could you share the code that you use to build this stalked barplot on intersect area??

**Comment by efr3m** [comment]
Is there any updated on this issue? I would be really interested in this feature.

**Comment by radlinsky** [comment]
@jananiravi  @acpguedes I'm not really sure how to scale this up if you have more than two "lineages" as shown in @jananiravi's example. But here is my solution for a simpler case where each element in the set can either be UP or DOWN (for e.g. differentially expressed gene is up- or -down-regulated).

![Screen Shot 2019-03-21 at 1 32 23 PM](https://user-images.githubusercontent.com/5822595/54773183-6f621580-4bdf-11e9-8614-0acf3ebccad9.png)

![upset_direction_included](https://user-images.githubusercontent.com/5822595/54773173-68d39e00-4bdf-11e9-8e41-a18b7e983bed.png)
 
[upset_plot_with_directionality.txt](https://github.com/hms-dbmi/UpSetR/files/2993326/upset_plot_with_directionality.txt)

**Comment by janstrauss1** [comment]
> @jananiravi @acpguedes I'm not really sure how to scale this up if you have more than two "lineages" as shown in @jananiravi's example. But here is my solution for a simpler case where each element in the set can either be UP or DOWN (for e.g. differentially expressed gene is up- or -down-regulated). The one "caveat" (or advantage, depending on your point of view) is that any given gene could be counted twice... once for being UP and a second time for being DOWN...
> 
> ![Screen Shot 2019-03-21 at 1 32 23 PM](https://user-images.githubusercontent.com/5822595/54773183-6f621580-4bdf-11e9-8614-0acf3ebccad9.png)
> 
> ![upset_direction_included](https://user-images.githubusercontent.com/5822595/54773173-68d39e00-4bdf-11e9-8e41-a18b7e983bed.png)
> 
> [upset_plot_with_directionality.txt](https://github.com/hms-dbmi/UpSetR/files/2993326/upset_plot_with_directionality.txt)

@radlinsky, I am still unclear how you generate the stacked bar chart. Could you please share your code to reproduce your example?

Many thanks in advance!

**Comment by radlinsky** [comment]
Hey @janstrauss1, I thought I added a txt file with the code to reproduce the plot, but maybe it didn't attach. Here:

`
library(UpSetR)
test <- as.data.frame(matrix(rnorm(1:12), nrow = 4))
colnames(test) <- c("A", "B", "C")
test_up <- test
test_up[test >= 0] <- 1
test_up[test_up != 1] <- 0
test_up$Gene <- paste0("Gene", 1:nrow(test), "_UP")
test_down <- test
test_down[test < 0] <- 1
test_down[test_down != 1] <- 0
test_down$Gene <- paste0("Gene", 1:nrow(test), "_DOWN")
test <- rbind(test_up, test_down)


upordown <- function(row, direction) {
  gene <- row["Gene"]
  if (grepl(x = gene, pattern = direction)) {
    newData <- T
  }
  else {
    newData <- F
  }
}


metadata <- data.frame(
  c("A", "B", "C"),
  as.numeric(apply(test_up[, 1:3], 2, sum))
)
colnames(metadata) <- c(
  "sets",
  "NumberUP"
)


upset(test,
      sets = c("A", "B", "C"), set.metadata = list(
        data = metadata,
        plots = list(
          list(type = "hist", 
               column = "NumberUP", 
               assign = 20, # defines width of the meta-data histogram
               colors = "red")
        )
      ),
      queries = list(list(query = upordown, params = list("_UP"), color = "red", active = TRUE))
)
`

**Comment by greboul** [comment]
Sadly it looks like this project has been put on hold....

But here is my contribution to this possible stacked barplot feature to add to upsetr because i think it can be very useful to gather closely related informations on the same figure (for a scientific article).

Attached are the barplot and the upsetr plot I would have liked to combine easily and not using Inkscape manually to copy paste each barplot on top of the corresponding regroupments. Using this combined plot it is easier to understand and vizualize which samples share common feature and it can be easily be related to the composition of each.

1/ barplot alone
![mexbarplot](https://user-images.githubusercontent.com/13840904/57277304-289f8080-70a4-11e9-9111-840c95799c2e.gif)

2/ upsetr plot alone
![mexbarplot2](https://user-images.githubusercontent.com/13840904/57277423-6dc3b280-70a4-11e9-9d8f-8f4d8a72d20f.gif)

3/ both combined 
![mexbarplot3](https://user-images.githubusercontent.com/13840904/57278148-06a6fd80-70a6-11e9-870b-d8f930696e46.gif)

Hope this does help to understand why it could be a really nice add-on to this package and what I (we) would expect to do then !

Thanks,
Greboul

**Comment by dlaehnemann** [comment]
This question also arose on stackoverflow, and I found a workaround for the main (intersection) bars that I posted there:
https://stackoverflow.com/a/56704255/2352071

What remains impossible with that workaround is overlaying the set sizes with the query annotation. This would be another nice feature to have.

**Comment by samlipworth** [comment]
+ 1 for @dlaehnemann's suggestion - if this could be native it would be great. @dlaehnemann your stack answer is great - anyway to add a legend to show what the colours in the stacked barchart mean?

**Comment by dlaehnemann** [comment]
Sorry, I just resorted to creating that manually, as I only had a single plot.

But `Example 5` in this tutorial might work:
https://cran.r-project.org/web/packages/UpSetR/vignettes/queries.html

Otherwise, there seems to be a recent reimplementation based on the grammar of graphics, that should allow for easier customization. Not sure if it supports as many panels in the grid, though:
https://github.com/const-ae/ggupset

**Comment by samlipworth** [comment]
thanks those are both really helpful suggestions - Example 5 does indeed work.

**Comment by krassowski** [comment]
I just saw this in my e-mail. I have a package which re-implements upset called [ComplexUpset](https://github.com/krassowski/complex-upset/) using ggplot2. Would the following work for you: 

![image](https://user-images.githubusercontent.com/5832902/83152331-9aa29780-a0f5-11ea-99e1-60eee1169efb.png)

https://krassowski.github.io/complex-upset/articles/Examples_R.html#fill-the-bars

The code is slightly different but the gist is:

```R
upset(
    movies,
    genres,
    base_annotations=list(
        'Intersection size'=intersection_size(
            counts=FALSE,
            aes=aes(fill=mpaa)
        )
    ),
    width_ratio=0.1
)
```

**Comment by dlaehnemann** [comment]
Nice, you might also want to add this info to the stackoverflow question, that does seem to get some traffic! :)

https://stackoverflow.com/a/56704255/2352071

**Comment by samlipworth** [comment]
@krassowski very neat, thanks

**Comment by gtrichard** [comment]
Very nice, I guess it closes the issue !
